### DISCUSSION TITLE
Help: Slow Multiple SVG Rendering (while creating String Lights effect)

### DISCUSSION THREAD

**Discussion by kartiksolanki** [fabricjs/fabric.js#8627] [Q&A]
Hello All,
I am trying to create Holiday Christmas Lighting (String Light) effect in Fabricjs. My string lights consist of repetative SVG icons (bulbs) which I am rendering as follow:
1. Creating a polyline based on points array.
2. using the same points array, I am calculating spacing in order to get the coordinates array of where each bulb inside string light will be placed. This spacing I am calculating based on bulb counts.
3. Then I am doing a for loop based on bulb counts and doing loadSVGfromString operation where I am adding each bulb (canvas.add) count on the polyline points based on spacing coordinates array that I have calculated in step 2.
4. When the count is last in the loop, I am doing canvas.renderAll().

Using the above logic I am able to successfully create string light demonstration on the canvas but the problem is, it takes few milliseconds to render each bulb on the each spacing coordinates. From my belief, fabric js is crazy fast in rendering so I am wondering, if I am doing anything wrong or is there any optimized way to create string light effect instantaneously ??? Please guide me.

**Comment by iocoker** [comment]
`loadSVGfromString` is pretty expensive. If the bulbs look the same, have you tried creating a single bulb first, and then clone that first bulb instead?

Alternatively, If the bulbs are all different, you may try rendering them all `(loadSVGfromString)` in a `Promise.all` before starting to place them on the polyline.

**Reply by kartiksolanki** [reply]
I am not sure how would I make clone work since I am using index inside the for loop to provide top, left value as well as multi color bulbs based on index and multicolor array. But I tried `Promise.all` and it did give me a minute improvement which is not bad.

**Reply by kartiksolanki** [reply]
Also after my for loop i am grouping individual bulbs into a single group so how do i make that possible with clone?? Can you give me a rough structure on how would you approach cloning with string lights.

**Comment by ShaMan123** [comment]
I suggest you add a repro to move forward.
@iocoker suggests good advice, I am sure that if you provide a working example it will take a few mins to suggest a solution.

**Comment by asturur** [comment]
Definetely a reproducible bulb + string light in a fiddle would make this a fun discussion where everyone can get creative with suggestions.
Depending on how the bulb is complex, and how much it has to vary, there may be different solutions, one could also be load it as an image.

**Reply by kartiksolanki** [reply]
I thought about image but in my use-case SVG fits getter as I will be providing dynamic feature to change light bulbs color. So I have created custom SVG with path names such that the fabricjs recognizes those path and I am able to change the fill and gradient. I can do image but it doesn't make sense to have bulb images of different colors and render the specific colored bulb img based on selection. With SVG, I can just utilize awesome fabricjs paths feature and change fill and gradient easily. So it has to be SVG.

**Comment by kartiksolanki** [comment]
Sounds good. I will make a fiddle this weekend and share it with you guys.
